### ISSUE TITLE
[Epic]: Blockify Add to Cart with Options block

### ISSUE THREAD

**Issue by octaedro** [woothemes/woocommerce#53014]
### Description

This epic encompasses all tasks related to blockifying the `Add to Cart Form`.

* [GitHub Board 1](https://github.com/orgs/woocommerce/projects/284)
* [GitHub Board 2](https://github.com/orgs/woocommerce/projects/294)
* [GitHub Board 3](https://github.com/orgs/woocommerce/projects/305)

**Comment by helgatheviking** [comment]
Following!

**Comment by helgatheviking** [comment]
Would you want to add asynchronous add to cart? I think you'd need to add support to the [`addToCart` function](https://github.com/woocommerce/woocommerce/blob/25fb13a2bf44fe92fef2e764adee53899422cb63/plugins/woocommerce-blocks/assets/js/data/cart/thunks.ts#L204-L214) for more parameters. Other product types will have more fields than ID and Quantity. I opened https://github.com/woocommerce/woocommerce/issues/53357 about this.

I would also like to suggest that the form have interactivity that includes some kind of validation.... so plugins can tell the form it's not valid and therefore should not be add to the cart. This will be needed by variable products anyway so it would be handy to have it accessible from other plugins from start. For example, in Name Your Price the add to cart button prevents submitting the form if the entered price is below the minimum. Same would be needed for Product Addons that are required... etc.

**Comment by Aljullu** [comment]
As usual, thanks for the feedback, @helgatheviking!

> Would you want to add asynchronous add to cart? I think you'd need to add support to the [addToCart function](https://github.com/woocommerce/woocommerce/blob/25fb13a2bf44fe92fef2e764adee53899422cb63/plugins/woocommerce-blocks/assets/js/data/cart/thunks.ts#L204-L214) for more parameters. Other product types will have more fields than ID and Quantity. I opened https://github.com/woocommerce/woocommerce/issues/53357 about this.

I think internally we will likely use the same Store API endpoint to handle Add to Cart events, but via the Interactivity API.

For 3rd-party extensions that only register some fields in the form, I hope we can make this transparent so they don't need to hook into that function or anything similar. IMO we should ideal detect the `name` property of every input and send it with the request.

For 3rd-party extensions that create new product types, I think it's still too early to say how it will look like. I still expect the wrapper block will be shared and hopefully that block will handle most of the logic of Adding to Cart. But I expect we will need to work on a few extensions until we find a flexible API that doesn't require extensions to rewrite the same logic over and over again.

> I would also like to suggest that the form have interactivity that includes some kind of validation.... so plugins can tell the form it's not valid and therefore should not be add to the cart. This will be needed by variable products anyway so it would be handy to have it accessible from other plugins from start. For example, in Name Your Price the add to cart button prevents submitting the form if the entered price is below the minimum. Same would be needed for Product Addons that are required... etc.

Good point. I expect there will be some validation on the browser side (ie: required fields, number inputs not allowing letters, etc.) and some on the server side (ie: complex rules like if Product A is in cart, Product B can't be added). So if a shopper inputs invalid data, the Add to Cart block will fail as expected. Is that aligned with what you had in mind, @helgatheviking?

**Comment by helgatheviking** [comment]
>IMO we should ideal detect the name property of every input and send it with the request.

Yes!! I second this. I think that as long as you also handle `quantity[99]` type of names (like in Grouped Products and Mix and Match products) this will cover the majority of cases. Something like this maybe?

```
yield wp.data.dispatch( cartStoreKey ).addItemToCart(
					productId,
					quantityToAdd,
					additionalData
				);
```

[Though the `cart-add-item` endpoint is expecting `variation` in a specific format](https://github.com/woocommerce/woocommerce/blob/53a2f97932fbdb81c404081cfd906f0fdd2f2140/plugins/woocommerce/src/StoreApi/Routes/V1/CartAddItem.php#L121).

>Good point. I expect there will be some validation on the browser side (ie: required fields, number inputs not allowing letters, etc.) and some on the server side (ie: complex rules like if Product A is in cart, Product B can't be added). So if a shopper inputs invalid data, the Add to Cart block will fail as expected. Is that aligned with what you had in mind,

Yes. Another example would be in Mix and Match products I require customers to select X quantities of products before you can add the pack to the cart. Similar functionality in Bundles. If not valid, we show a notice/message.

I was going to play around with adding that to the interactivity store that was [stubbed out here](
https://github.com/woocommerce/woocommerce/pull/53215/files#diff-a0064c1e28a09f52947e6d5be8c43730b54d077082989448215c3e455def0c3e) but I think there are some build and enqueuing issues that need to be worked out first (to use the iAPI you have to enqueue script modules and use experimental features of the WordPress build scripts so I don't know how y'all will make that work here)

I have something brewing in my own extension of the Add to Cart With Options block and core iAPI that I will share when it's a bit cleaner. At the moment, it's a lot easier for me to work in my own plugin than using the woo mono-repo tooling.

**Comment by helgatheviking** [comment]
https://www.loom.com/share/afa7ba1f4d914fb98ca1c69d15fbf9fb

Needs some animation, but I have modified `addItemToCart()` to support at least the variable form. Still working on supporting additional fields.

**Comment by Aljullu** [comment]
> Yes. Another example would be in Mix and Match products I require customers to select X quantities of products before you can add the pack to the cart. Similar functionality in Bundles. If not valid, we show a notice/message.

Makes sense. I think which kind of validation is done will need to be decided on an extension-per-extension basis. We will need validation on the server side for sure. And if we can add validation on the frontend side, that would be better. It's not yet clear how we will tackle that for complex validations besides min/max quantities, required fields, etc.

> https://www.loom.com/share/afa7ba1f4d914fb98ca1c69d15fbf9fb
> 
> Needs some animation, but I have modified `addItemToCart()` to support at least the variable form. Still working on supporting additional fields.

Awesome! Is that using inner blocks or is the current _Add to Cart with Options_ block using a JS-approach?

**Comment by catthetech** [comment]
When I try change the product type to from `Simple product` `Variable product` in the `Add to Cart with Options (Experimental)` block's toolbar (the blockified version), save, then reload the editor, my changes were not saved.
Is this feature implemented?, or does anyone know why it doesn't work?

**Comment by Aljullu** [comment]
> When I try change the product type to from `Simple product` `Variable product` in the `Add to Cart with Options (Experimental)` block's toolbar (the blockified version), save, then reload the editor, my changes were not saved.
Is this feature implemented?, or does anyone know why it doesn't work?

I'm not able to reproduce, I can modify the inner blocks as expected (ie, change their order). What exactly are you modifying, @catthetech?

**Comment by helgatheviking** [comment]
@Aljullu is there a way to register inner blocks for a specific product type yet? Like could I start playing around with it? 

Also what about a "type" like Name Your Price inputs which apply to multiple product types?

**Comment by Aljullu** [comment]
> is there a way to register inner blocks for a specific product type yet? Like could I start playing around with it?

Not yet, we are planning to work on this soon, but right now there is only one form shared by all product types, so there is no way to enable inner blocks only for specific product types.

What this selector does now is to provide context to the inner blocks, so they can change their appearance depending on the product type, but the form is still the same no matter the selected product type.

> Also what about a "type" like Name Your Price inputs which apply to multiple product types?

The selector displays the product types returned by `wc_get_product_types()`. My understanding is that _Name Your Price_ is not a product type itself, but a setting that can be turned on in several product types.

In that case, I would expect the _Name Your Price_ block to always be visible in the editor when editing generic templates. When editing the _Single Product_ block, as there is the product context, it should be possible to hide it or show a placeholder mentioning whether that block will be visible or not.

I would think of it as the _Product Rating_ block, which is always visible in the editor, but in the frontend it only renders when relevant. Do you think that makes sense?

**Comment by helgatheviking** [comment]
Thanks Albert! I am working on building out my inner block right now for Mix and Match (hopefully using iAPI), so I'll just wait on the registration part and if we're lucky it will all come together.

Name Your Price is a setting on _some_ supported product types...

**Comment by catthetech** [comment]
> I'm not able to reproduce, I can modify the inner blocks as expected (ie, change their order). What exactly are you modifying, [@catthetech](https://github.com/catthetech)?

I am able to modify the inner blocks as expected.
My issue is concerning the product type in the block's toolbar (right above the `Add to Cart with Options (Experimental)` block, or at the top of the screen), where it says `Simple product`, when I change this to anything else, say, `Variable product`, save then reload the editor, this specific change is not saved.

**Comment by Aljullu** [comment]
> My issue is concerning the product type in the block's toolbar (right above the `Add to Cart with Options (Experimental)` block, or at the top of the screen), where it says `Simple product`, when I change this to anything else, say, `Variable product`, save then reload the editor, this specific change is not saved.

Hmmm, right, the product type selector lets you see how the template will be rendered for each product type, but it's not something saved to the template itself. So the template is a generic _Single Product_ template that is used for all products, but you can change the visible product type with the selector to see how the form (and other blocks) will look like in that product type. That's why when you reload the page, you get the _Simple product_ type active again. Does it make sense? I expect once the different layouts per product type are done, it will be clearer how it works.

**Comment by catthetech** [comment]
Yeah that makes sense, thanks for the clarification on how this feature works.

**Comment by Aljullu** [comment]
I wanted to publish a project update for everybody following this thread and wanting to catch-up on the project status.

### Inner blocks

Inner blocks are mostly ready. We are planning to rewrite the _Variation Selector_ from scratch with a new design based on swatches/pills, but besides that, all the other inner blocks are done. Those include:

* Quantity Selector
* Add to Cart Button (already existed)
* Stock Notice (already existed)
* Variation Selector
* Grouped Product Selector

We still have some small issues here and there for some of them, but the big chunk of work is done.

### Handling different blocks per product type

The logic so we have a different template part per product type is also there. It allows switching the 'view' of the product type in the editor, and it handles rendering the correct template part in the frontend. It also handles 3rd-party registered product types, by displaying a placeholder in the editor and rendering the correct template in the frontend.

> [!NOTE]
>
> You might have noticed in the [initial plan](https://github.com/woocommerce/woocommerce/discussions/50484) we spoke about _Synced patterns_ instead of _Template parts_. The fact is that we found some limitations in the way WordPress currently handles synced patterns registered by plugins. That meant that they were not a valid option for our use case as we initially thought. Instead, template parts seem to meet all our requirements.

### Next steps

In the following weeks, we will start exploring how the iAPI will handle interactions with all the inner blocks and what extensibility will look like.

### Extensibility

Speaking about extensibility, what's ready and what's not?

Extensibility type | Current status
--- | ---
Themes overriding default template parts | ✅ Ready for testing. Themes can add `parts/<product_type>-add-to-cart-with-options.html` files to override the default WooCommerce template parts.
Extensions registering custom product types | ✅ Ready for testing. Custom product types are selectable from the Add to Cart with Options block product type switcher, they display a placeholder in the editor and the correct template in the frontend.
Extensions hooking into PHP hooks used in the form | 🛑 Not ready for testing. We are currently exploring how many of the existing PHP hooks can be supported via a compatibility layer. That involves not only rendering the contents but also handling form elements via the iAPI.
Extensions modifying the default product types forms | 🛑 Not ready for testing. Extensions that hook into `woocommerce_<product_type>_add_to_cart` are not supported yet.
Extensions wanting to switch to a blockified paradigm | 🛑 Not ready for testing. We haven't introduced any new extensibility points for now, so it's not possible to start updating existing extensions to the new blockified paradigm.

Please keep in mind _Ready for testing_ doesn't mean _Ready for production_, this project is still in active development and it's likely there will be many changes before it's released. But if you are curious or would like to take an early look at how your theme/extension might look like, you can do so! To test it, you will need to install and activate the WooCommerce Beta Tester plugin and go to _WooCommerce Admin Test Helper_ > _Features_ and enable `experimental-blocks` and `blockified-add-to-cart`.

**Comment by Aljullu** [comment]
Let me publish a quick update on where we are at this moment, as some time has happened since the [previous update](https://github.com/woocommerce/woocommerce/issues/53014#issuecomment-2637680874):

### Inner blocks

Most inner blocks are ready. Those include:

* Quantity Selector
* Variation Selector and all its inner blocks, including 'pill' and 'dropdown' styles
* Grouped Product Selector and all its inner blocks

We are still getting design feedback on how the blocks should behave once a variation is selected. That might require creating a few more inner blocks.

### Interactivity API

We have synced with some engineers with more knowledge of the Interactivity API to plan the next steps for the Add to Cart with Options block. So far, the pills-based Variation Selector already uses the Interactivity API (https://github.com/woocommerce/woocommerce/pull/56140), and we recently merged the PR (https://github.com/woocommerce/woocommerce/pull/56378) that allows adding variable products to cart. We still need to add support for simple and grouped products, quantities, etc. but progress is being made in this field.

### Styling updates

As mentioned above, we implemented the new pill-based _Variation Selector_. @amitraj2203 also helped making several updates to the inner blocks, so they work well in a variety of themes and color combinations (thanks! :partying_face: ).

Before | After
--- | ---
![imatge](https://github.com/user-attachments/assets/d8df3222-b2d2-4763-a603-94f3fe59fda9) | ![imatge(1)](https://github.com/user-attachments/assets/6552463b-18fb-417a-b6f2-51125659bd44)

### Next steps

The next steps consist:

1. Finish implementing the iAPI-powered add-to-cart behavior for all product types.
2. Finalize backwards-compatibility checks with current extensions.
3. (Maybe) introduce some changes to the behavior when selecting variations and to the added-to-cart notice.
4. Work on block-based extensibility points.

### Extensibility

Speaking about extensibility, what's ready and what's not?

Extensibility type | Current status
--- | ---
Themes overriding default template parts | ✅ Ready for testing. Themes can add `parts/<product_type>-add-to-cart-with-options.html` files to override the default WooCommerce template parts.
Extensions registering custom product types | ✅ Ready for testing. Custom product types are selectable from the Add to Cart with Options block product type switcher, they display a placeholder in the editor and the correct template in the frontend.
Extensions hooking into PHP hooks used in the form | 🛑 Not ready for testing. We will likely "fall back" to a regular `<form>` when there are extensions hooking into the PHP hooks of the form.
Extensions modifying the default product types forms | ✅ Ready for testing. Extensions using the `woocommerce_<product_type>_add_to_cart` hook on WC core product types are already supported (https://github.com/woocommerce/woocommerce/pull/56536).
Extensions registering custom product types wanting to switch to a blockified paradigm | 🛑 Not ready for testing. @helgatheviking has submitted a PR (https://github.com/woocommerce/woocommerce/pull/56799) that will allow extensions to register their own template parts to hook in the form (thanks! :partying_face: ).
Extensions hooking into PHP hooks used in the form wanting to switch to a blockified paradigm | 🛑 Not ready for testing. We are brainstorming several ways of achieving this that would involve using Block Hooks and the Interactivity API

Please keep in mind we aren't targeting 100% backwards compatibility with existing extensions because it's impossible. The compatibility layer that brings support for classic PHP hooks is there to ease the transition for existing extensions until we have the block-based extensibility points in place.

Also, keep in mind that _Ready for testing_ doesn't mean _Ready for production_, this project is still in active development and it's likely there will be many changes before it's released. But if you are curious or would like to take an early look at how your theme/extension might look like, you can do so! To test it, you will need to install and activate the WooCommerce Beta Tester plugin and go to _WooCommerce Admin Test Helper_ > _Features_ and enable `experimental-blocks` and `blockified-add-to-cart`.

If you are interested, in order to stay up to date about this project, I would suggest subscribing to this issue, [this discussion](https://github.com/woocommerce/woocommerce/discussions/50484) and following the [`add to cart form` label](https://github.com/woocommerce/woocommerce/issues?q=state%3Aopen%20label%3A%22block%3A%20add%20to%20cart%20form%22) in the repo.

**Comment by 79ho3ein** [comment]
Hi woocommerce team,

Is it currently possible to test and review the new Add to Cart block?

**Comment by Aljullu** [comment]
Hi there, @79ho3ein, thanks for your interest! The block is still in active development and not ready to use in live stores. Having said that, you can test it by following these steps:

1. Install and activate the [WooCommerce Beta Tester plugin](https://woocommerce.com/products/woocommerce-beta-tester/).
2. Go to Tools WA Test Helper > Features and enable `experimental-blocks` and `blockified-add-to-cart`.
3. Now, when you edit the Single Product template or the Single Product block and click on the _Add to Cart with Options_ block, you will see an option to upgrade in the sidebar.

Hope that helps!

**Comment by 79ho3ein** [comment]
Thanks for the great advice @Aljullu

**Comment by Aljullu** [comment]
Publishing a quick update as it has been more than two months since the last one:

### Intro

The _Add to Cart + Options_ block is a new block intended to be used as a form to add products to cart, usually in the _Single Product_ template or inside the _Product_ block. It allows a much higher degree of customization via its inner blocks, as well as bringing several quality of life for shoppers, like adding to cart without a page refresh.

![Image](https://github.com/user-attachments/assets/deb77c1a-ee73-4467-9bd0-9bf298c6e94c)

### Highlights

* The block will be released as a beta in WooCommerce 10.0. 🥳 
* We renamed it from _Add to Cart with Options_ to _Add to Cart + Options_. We renamed many of the inner blocks and their descriptions as well.
* Some block slugs changed too. That's important in case you had updated or override one of the _Add to Cart + Options_ template parts.
* All the important functionality for adding to cart is already in place. The next steps will be extensibility and quality of life improvements.

### Inner blocks

All inner blocks have been completed. Since the last update, we improved several of the existing blocks, and introduced a new one named _Grouped Product: Item Label_, which is used to display the product name of grouped product children. This block differs from the previous behavior in that it's now a `<label>` attached to the quantity input, instead of being a link to the children product page. That was a decision we took for accessibility reasons.

![Image](https://github.com/user-attachments/assets/2a359070-9d0e-4726-9f48-f34634d034b5)

### Interactivity API

Adding to cart with the Interactivity API is fully supported in all WooCommerce core product types, That means, adding to cart without a page refresh! 🎉

We are currently working on making some blocks update when a specific variation is selected. In WooCommerce 10.0, only the _Product Price_ block has this functionality, but the plan is to expand it to other blocks (https://github.com/woocommerce/woocommerce/issues/56870) and eventually let extensions and 3rd-party blocks do the same.

We are also planning a big refactor of the Interactivity API stores used by the _Add to Cart + Options_ and other blocks. The idea is that all data which is the same across blocks, will be in one shared store, instead of each block having its own copy of the data in their store. This will avoid repetition and make it easier for blocks consuming the data to have a single source of truth. This is still in a planning phase, but mentioning it because the Interactivity API stores will likely change between now and when the block is marked as stable.

### Next steps

The next steps consist:

1. Fine-tune the existing experience: we are aware of several quality of life improvements we can introduce to the _Add to Cart + Options_ block and we will focus on that in the following weeks (you can check issues with the [`add to cart form` label](https://github.com/woocommerce/woocommerce/issues?q=state%3Aopen%20label%3A%22block%3A%20add%20to%20cart%20form%22) for some examples).
2. Refactor Interactivity API stores as mentioned above.
3. Work on block-based extensibility points.
4. Remove the beta label from the block and mark the existing _Add to Cart with Options_ block as _Classic_ or _Legacy_, eventually hiding it from the inserter.

### Extensibility

About extensibility, this is what's ready for testing:

Extensibility type | Current status
--- | ---
Themes overriding default template parts | ✅ Ready for testing. Themes can add `parts/<product_type>-add-to-cart-with-options.html` files to override the default WooCommerce template parts.
Extensions registering custom product types | ✅ Ready for testing. Custom product types are selectable from the _Add to Cart + Options_ block product type switcher, they display a placeholder in the editor and the correct template in the frontend.
Extensions hooking into PHP hooks used in the form | ✅ Ready for testing. We can't support _all_ PHP hooks, but we support several of them and when we detect they are being used by an extension, we "fall back" to a regular form that causes the form to reload.
Extensions modifying the default product types forms | ✅ Ready for testing. Extensions using the `woocommerce_<product_type>_add_to_cart` hook on WC core product types are already supported (https://github.com/woocommerce/woocommerce/pull/56536).
Express Payment Methods | ✅ Ready for testing. We updated several inner blocks to have similar classes and markup as the old templates, this makes many Express Payment Methods work out-of-the-box in simple products. For variable products, it's likely payment methods will need to be updated, as the UI has changed and now there is the chips/pills style.
Extensions registering custom product types wanting to switch to a blockified paradigm | 🛑 Not ready for testing. This is being tracked in https://github.com/woocommerce/woocommerce/issues/55700.
Extensions hooking into PHP hooks used in the form wanting to switch to a blockified paradigm | 🛑 Not ready for testing. This is being tracked in https://github.com/woocommerce/woocommerce/issues/55701.

Please keep in mind we aren't targeting 100% backwards compatibility with existing extensions because it's impossible. The compatibility layer that brings support for classic PHP hooks is there to ease the transition for existing extensions until we have the block-based extensibility points in place.

If you encounter any issues with elements marked as _Ready for testing_ and you think it's something that should be fixed in the _Add to Cart + Options_ block, please let us know creating a new issue or leaving a comment here.

And if you want to stay up to date with this project, you can subscribe to this issue, [this discussion](https://github.com/woocommerce/woocommerce/discussions/50484) and follow the [`add to cart form` label](https://github.com/woocommerce/woocommerce/issues?q=state%3Aopen%20label%3A%22block%3A%20add%20to%20cart%20form%22) in the repo.

**Comment by 79ho3ein** [comment]
Hi @Aljullu,

Thank you for all the great work. I had a question about the Add to Cart block...

Is it possible to place the variation selector and the add to cart button far apart? Or should they all be together?

Let me give you an example.

Please see the screenshot below. In this website, which is not built with WooCommerce, the variation selector is at the top of the product page, while the add to cart button is floating at the bottom and the price is displayed next to it. Is it possible to implement this with the new add to cart block?

If you don't understand what I mean, please view [this link](https://www.digikala.com/product/dkp-17986495/%DA%AF%D9%88%D8%B4%DB%8C-%D9%85%D9%88%D8%A8%D8%A7%DB%8C%D9%84-%D8%A7%D9%BE%D9%84-%D9%85%D8%AF%D9%84-iphone-16-ch-%D8%AF%D9%88-%D8%B3%DB%8C%D9%85-%DA%A9%D8%A7%D8%B1%D8%AA-%D8%B8%D8%B1%D9%81%DB%8C%D8%AA-128-%DA%AF%DB%8C%DA%AF%D8%A7%D8%A8%D8%A7%DB%8C%D8%AA-%D9%88-%D8%B1%D9%85-8-%DA%AF%DB%8C%DA%AF%D8%A7%D8%A8%D8%A7%DB%8C%D8%AA/) on your mobile. Although it is in Persian, it is understandable.

![Image](https://github.com/user-attachments/assets/d4e16f26-edf4-4171-af24-1fbdff34da4c)

**Comment by Aljullu** [comment]
Good question, @79ho3ein! This layout is something we evaluated at the beginning of the project but decided to leave for a later stage. Right now if you add two _Add to Cart + Options_ block to the same page, they are not in sync. This is relatively easy to solve, but there is the extra challenge of accessibility: the variation selector and the _Add to Cart Button_ need to belong to the same `<form>`. I opened an issue to investigate that in https://github.com/woocommerce/woocommerce/issues/59208. Thanks for your feedback! 🙌

**Comment by Joliamendi** [comment]
Wow, I can't wait to test the new add to cart block.


Has the Ajax limitation of variable products been resolved in the new version? Or still should we use the following code?


function custom_wc_ajax_variation_threshold( $qty, $product ) {
	return 100;
}

add_filter( 'woocommerce_ajax_variation_threshold', 'custom_wc_ajax_variation_threshold', 10, 2 );

**Comment by Aljullu** [comment]
@Joliamendi good question! Unfortunately I don't have a definitive answer yet. We haven't set a specific limit of variations that can be loaded in the new _Add to Cart + Options_ block yet, but it's likely we won't be able to load an undefined number of them without an upper limit for performance reasons.

I agree that `woocommerce_ajax_variation_threshold` being 30 by default feels a bit too small, nowadays, so I would like to explore ways for the new block to have a higher default. But I don't know yet how it will be done and whether we will use the same filter or not. We will likely investigate it as part of this issue: https://github.com/woocommerce/woocommerce/issues/57727.

**Comment by Aljullu** [comment]
I will go ahead and close this epic. The _Add to Cart + Options_ block has been available as a beta since WooCommerce 10.0. We are continuing to improve it and beginning to work on the remaining extensibility needs, but this epic was specifically aimed at getting the block released, and that goal has already been met. 🙂
